keps/sig-cluster-lifecycle/kubeadm/20190506-kubeadm-machine-output.md

```diff
@@ -0,0 +1,370 @@
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Kubeadm should support structured output, such as JSON, YAML, or a [Go template](https://golang.org/pkg/text/template/).
+
+## Motivation
+
+While Kubernetes may be deployed manually, the de facto, if not de jeur, means of turning up a Kubernetes cluster is kubeadm. Popular systems management software, such as Terraform, rely on kubeadm in order to deploy Kubernetes. Planned enhancements to the Cluster API project include a composable package for bootstrapping Kubernetes with kubeadm and cloud-init.
+
+Without structured output, even the most seemingly innocuous changes could break Terraform, Cluster API, or other software that relies on the results of kubeadm.
+
+### Goals
+
+* Kubeadm should support structured output, including, but not limited to, the following output types:
+  * Buffered
+    * JSON
+    * YAML
+    * Go template
+  * Unbuffered
+    * Text (the current behavior)
+  * Please see [*Buffered vs Unbuffered*](#buffered-vs-unbuffered) for details on what both terms mean in the context of this KEP
+* Kubeadm should support the above output types for commands that include, but not necessarily limited to, the following:
+  * `alpha certs`
```

Comment:
Can the parsable output for all these commands be a single omit-empty stubbed runtime-type struct?

Reply:
Hi @timothysc,

Maybe. Could also be a twinkie. Or a '57 Chevy Camaro. Either way, I think this is a duplicate of your other comment - https://github.com/kubernetes/enhancements/pull/1054/files#r282585029. If it's okay with you, I'd like to mark this comment resolved and use the other comment where I already responded with a non-stupid remark :)

Reply:
see other comment